src/Cms/AppPlugins.php

```diff
@@ -225,6 +227,63 @@ protected function extendFileMethods(array $methods): array
         return $this->extensions['fileMethods'] = File::$methods = array_merge(File::$methods, $methods);
     }
 
+    /**
+     * Registers additional custom file types and mimes
+     *
+     * @param array $fileTypes
+     * @return array
+     */
+    protected function extendFileTypes(array $fileTypes): array
+    {
+        $extends = [
+            'type' => [],
+            'mime' => [],
+            'resizable' => [],
+            'viewable' => []
+        ];
+
+        // normalize array
+        foreach ($fileTypes as $ext => $file) {
+            $extension = $file['extension'] ?? $ext;
+
+            if (isset($file['type']) === true) {
+                $extends['type'][$file['type']][] = $extension;
+            }
+
+            if (isset($file['mime']) === true) {
+                $extends['mime'][$extension] = $file['mime'];
+            }
+
+            if (($file['resizable'] ?? false) === true) {
+                $extends['resizable'][] = $extension;
+            }
+
+            if (($file['viewable'] ?? false) === true) {
+                $extends['viewable'][] = $extension;
+            }
+        }
+
+        // inject extended extensions
+        foreach ($extends as $type => &$values) {
+            switch ($type) {
+                case 'type':
+                    $values = F::$types = array_merge_recursive(F::$types, $values);
```

Comment:
Who has better suggestions than `array_merge_recursive()`?

Reply:
I think it's fine for the purpose. An alternative could be `array_replace_recursive()`, but it does the same for our array structure as far as I remember.

Do you see any downsides to using this function?

Another alternative could be to restructure the method like this:

- The `$extends` variable at the beginning is removed
- Instead, the loop through `$fileTypes` updates the properties of the different classes directly
- The `$extends` variable is then constructed *afterwards* with the values from the class properties

This would also allow us to get rid of the second `foreach` with `switch` that is currently a bit hard to understand.

Reply:
> Do you see any downsides to using this function?

I had a doubt in my mind but false alarm, it's fine. But as you said, I think it is more correct to use `array_replace_recursive()` to avoid duplicate entry.

> Instead, the loop through $fileTypes updates the properties of the different classes directly

You mean like following?

````php
// normalize array
foreach ($fileTypes as $ext => $file) {
    $extension = $file['extension'] ?? $ext;

    if (isset($file['type']) === true) {
        // inject to `F::$types` each fileType entry ??
        F::$types = array_replace_recursive(F::$types, [$file['type'] => [$extension]]);
    }
    
    ...
    ...
    ...
}
````

As I have concern in above code;
Since using `array_merge()` each time while normalizing each type creates a performance problem, I first put it into a temporary array and then merged it. Am I wrong?

Reply:
Yep, pretty much. But wouldn't a simple `F::$types[$file['type']][] = $extension` work as well? The only downside would be that an extension could end up in the array multiple times. We can solve this by first checking if the extension is not already defined.

Reply:
Sometimes I overthink 😅

Reply:
How does it look now?